### PR TITLE
Add `-Stable` to `Sort-Object` and related tests

### LINKED ISSUE #7846: Sort-Object inappropriately sacrifices stability for performance
PR #2518 by @KirkMunro adds the ability to select the top or bottom `N` items of a sorted set, introducing stability to the sorting method. However, this behavior is undermined whenever `N` equals or exceeds the input size.

This deviance can be observed when sorting by congruence modulo for example:

Steps to reproduce
------------------

```powershell
1..20 |Sort-Object {$_ % 3} -Top 20
```

Expected behavior
-----------------

```none
3
6
9
12
15
18
1
4
7
10
13
16
19
2
5
8
11
14
17
20
```

Actual behavior
---------------

```none
18
3
15
6
12
9
1
16
13
10
7
4
19
11
8
14
5
17
2
20
```

The culprit seems to be the last two conditions in this branch: https://github.com/KirkMunro/PowerShell/blob/master/src/Microsoft.PowerShell.Commands.Utility/commands/utility/sort-object.cs#L247

```csharp
            // If -Top & -Bottom were not used, or if -Top or -Bottom would return all objects, invoke
            // an in-place full sort
            if ((Top == 0 && Bottom == 0) || Top >= dataToProcess.Count || Bottom >= dataToProcess.Count)
            {
                sortedItemCount = FullSort(dataToProcess, comparer);
            }
            // Otherwise, use an indexed min-/max-heap to perform an in-place sort of all objects
            else
            {
                sortedItemCount = Heapify(dataToProcess, comparer);
            }
```

Suggested fix would be to either add a `-Stable` switch parameter that forces `Sort-Object` to use the stable min/max heap sort implementation, or change the branch above to only default to `FullSort()` when `-Top`/`-Bottom` are omitted completely.

My personal preference would be to go for _both_ changes - `-Stable` would be nice for when you don't know input size up front but want to maintain input order for equal items

Environment data
----------------

```powershell
> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      6.1.0
PSEdition                      Core
GitCommitId                    6.1.0-57-g0f0c46dfe51440470b86f67b258e14aea32fa0aa
OS                             Microsoft Windows 10.0.17134 
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### PR DESCRIPTION
**Pull request by KirkMunro**
## PR Summary

Fix #7846.

- added -Stable switch parameter to Sort-Object
- moved -Top parameter from Default parameter set to Top parameter set and made it mandatory for that parameter set
- added new Pester tests for -Stable switch (ascending, descending, with/without unique)

Regarding the breaking change, this is clearly in [Bucket 3](https://github.com/PowerShell/PowerShell/blob/master/docs/dev-process/breaking-change-contract.md#bucket-3-unlikely-grey-area) (technically it is breaking, but it is highly unlikely that anyone will be impacted by this change). For details on how it is breaking, see [here](#issuecomment-424757855).

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [X] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [X] Issue filed - Issue link: https://github.com/PowerShell/PowerShell-Docs/issues/3246
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [X] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/sort-object.cs
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/sort-object.cs]: Please use auto-property.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Sort-Object.Tests.ps1
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Sort-Object.Tests.ps1]: We could use shorter range  `1..10`.

**Review by iSazonov — COMMENTED** [review]
Please look CodeFactor issues.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Sort-Object.Tests.ps1
Comment (KirkMunro) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Sort-Object.Tests.ps1]: Using a shorter range invalidates the test. A longer range is required for the test to be able to properly identify that the stable sort is working. There is no difference between a stable and an unstable short with a range of 1..10.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/sort-object.cs
Comment (KirkMunro) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/sort-object.cs]: My preference is absolutely for auto-properties; however, the reason I chose to use this format was simply to follow existing style for other SwitchParameter properties. My assumption is that whoever implemented sort had a reason for creating a boolean field to accompany a SwitchParameter property instead of using an auto-property. If we are to change this cmdlet to use auto-properties for parameters of type SwitchParameter, I think it should be done for all the switch parameters in the cmdlet, not just the one that I have added. Thoughts?

**Comment by KirkMunro** [comment]
@iSazonov Why did you add the Breaking-Change label? Because the Top parameter was moved to a separate parameter set (which would invalidate code someone had written specifically against the Default parameter set if they expected Top to be in there)? Or is there a functional breaking change you see here that I don't?

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/sort-object.cs
Comment (KirkMunro) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/sort-object.cs]: You know what, nevermind, I don't see any advantage to using a boolean field with a switchparameter parameter, so I made this change for both -Stable and -Unique parameters.

**Comment by iSazonov** [comment]
@KirkMunro Formally it is a breaking change because (1) a parameter set is changed, (2) sort order is changed and some scripts can fail. I believe we should inform users in change log about this. Also we need PowerShell Committee approvement.

**Comment by iSazonov** [comment]
@KirkMunro Please update Markdig.Signed package version 0.15.2 -> 0.15.3 to pass CI tests.

**Comment by anmenaga** [comment]
@SteveL-MSFT we need to make sure this breaking change gets reviewed by PS Committee...
`Review - Committee` label was added almost a month ago.

**Comment by SteveL-MSFT** [comment]
@anmenaga yes, this is in the queue for review but we have so many items now it's taking a bit longer

**Comment by iSazonov** [comment]
@KirkMunro Please resolve merge conflicts.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/sort-object.cs
Comment (PaulHigin) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/sort-object.cs]: Why did you remove `Top >= dataToProcess.Count || Bottom >= dataToProcess.Count` condition?

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/sort-object.cs
Comment (PaulHigin) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/sort-object.cs]: Please add to comment that Heap sort performs a stable sort, and provide short description of stable sort.

**Review by PaulHigin — CHANGES_REQUESTED** [review]

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/sort-object.cs
Comment (KirkMunro) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/sort-object.cs]: Done and done. Commit coming shortly.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/sort-object.cs
Comment (KirkMunro) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/sort-object.cs]: Done and coming in a commit shortly.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/sort-object.cs
Comment (KirkMunro) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/sort-object.cs]: The discussion in bug #7846 describes this in detail (and this was the original driver for this PR). With the old logic, when Top or Bottom exceed the number of items being sorted, you wouldn't end up with a heap sort, which you should when you use Top or Bottom. The end result was a less performant, unstable full sort when you expect a stable heap sort to be performed. Removing these conditions fixes that issue (one part of this PR). Adding -Stable addresses the need for stable sorting when you don't care about the original data size (the other part of this PR).

**Review by PaulHigin — APPROVED** [review]
LGTM

**Comment by SteveL-MSFT** [comment]
@BrucePay I believe you had an action item to review this PR on behalf of the @PowerShell/powershell-committee

**Review by adityapatwardhan — APPROVED** [review]

**Comment by adityapatwardhan** [comment]
Change looks good, awaiting review from committee.

**Comment by sdwheeler** [comment]
Docs issue https://github.com/PowerShell/PowerShell-Docs/issues/3021

**Comment by BrucePay** [comment]
@KirkMunro @anmenaga My main question is this: do we really need a custom implementation of a heap sort? LINQ OrderBy and OrderByDescrnding are stable sorts. Why not just use them? It would be a lot less code to maintain.

**Comment by iSazonov** [comment]
There has been a significant performance increase in # 2518 + support Top/Bottom.
There is open issues in CoreFX for new API that can replace the code.

**Comment by KirkMunro** [comment]
@BrucePay That question has nothing to do with this PR. This PR is to correct issue #7846, while also allowing full stable sorts using the same algorithm. The sorting algorithm isn't new here. It was added for PowerShell 6.0 in PR #2518 to greatly improve Sort-Object performance issues while adding -Top and -Bottom to Sort-Object.

To get the answer to your question, I think it would be better to open another issue requesting a comparison between LINQ's OrderBy and OrderByDescending and the existing indexed min-/max-heap sort algorithm, which should include an in depth analysis of results to ensure sorting does not change as well as a thorough performance analysis. Based on past experience investigating the performance issues with Sort-Object, and figuring out the algorithm to add -Top and -Bottom sort capabilities. Either that or compare with the CoreFX issue that @iSazonov just mentioned (Ilya, can you provide a link to that please?).

**Review by BrucePay — APPROVED** [review]
Hmm - sorry - I clicked the link from the issue and started looking at the wrong PR. Given that the code I'm concerned about has already been committed I agree that a separate issue is the right way to approach it. Other than that, you might consider checking PSBoundParameters to see if the top/bottom/stable parameters have been _specified_ instead of depending on signal values but I'll leave that up to you. (Sooner or later someone is probably going to suggest using `sort -Top -10` to get all _but_ the top ten :-) ) . Anyway I approve these changes.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/sort-object.cs
Comment (KirkMunro) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/sort-object.cs]: Fixed.

**Comment by iSazonov** [comment]
Waiting PowerShell Committee review.

**Comment by sdwheeler** [comment]
@SteveL-MSFT What's the status on this PR?

**Comment by adityapatwardhan** [comment]
@KirkMunro Fix the merge conflict.

**Comment by adityapatwardhan** [comment]
@KirkMunro Thank you for your contribution!

**Comment by KirkMunro** [comment]
Thanks @adityapatwardhan, for fixing the merge conflict and merging the PR. I missed the earlier comments about the merge conflict.

**Pull request merged**
